### PR TITLE
Volume is now usable once restoration is done.

### PR DESCRIPTION
**Pull request by ywei88**
Real restoration completed when restored volume first created, workload starting time has been reduced.

### CODE REVIEW

File: controller/volume_controller.go [hunk 1]
```diff
@@ -796,6 +811,13 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, e *longhorn
 		if oldState != v.Status.State {
 			vc.eventRecorder.Eventf(v, v1.EventTypeNormal, EventReasonAttached, "volume %v has been attached to %v", v.Name, v.Spec.NodeID)
 		}
+
+		//Automatically detach the new restored volume
+		if v.Spec.RestorationCompleted == false && v.Spec.FromBackup != "" && v.Spec.Standby == false {
```
Comment (shuo-wu) [inline controller/volume_controller.go]: it's better to add `v.Status.State == types.VolumeStateAttached` here.

File: controller/volume_controller.go [hunk 2]
```diff
@@ -528,14 +528,29 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, e *longhorn
...
 
+	allReplicaNew := true
+	for _, r := range rs {
+		if r.Spec.HealthyAt != "" {
+			allReplicaNew = false
+			break
+		}
+	}
+
+	if v.Spec.FromBackup != "" && v.Spec.Standby == false && allReplicaNew == true && v.Spec.RestorationCompleted == false {
```
Comment (shuo-wu) [inline controller/volume_controller.go]: Why do we need `allReplicaNew`? 
Can we use `v.Spec.FromBackup != "" && v.Spec.RestorationCompleted == false && v.Spec.Standby == false && v.Status.State ==  types.VolumeStateDettached` instead?

File: datastore/longhorn.go [hunk 3]
```diff
@@ -930,6 +930,31 @@ func (s *DataStore) ListNodes() (map[string]*longhorn.Node, error) {
...
+func (s *DataStore) GetRandomReadyNode() (*longhorn.Node, error) {
+	nodeList, err := s.ListNodes()
+	msg := fmt.Sprintf("nodelist %v", nodeList)
+	logrus.Errorf(msg)
+	if err != nil {
+		return nil, err
+	}
+	var usableNode *longhorn.Node
+	for name := range nodeList {
+		node, err := s.GetNode(name)
```
Comment (shuo-wu) [inline datastore/longhorn.go]: I think you can use the value of the map `nodeList` rather than `s.GetNode()`

File: datastore/longhorn.go [hunk 4]
```diff
@@ -930,6 +930,31 @@ func (s *DataStore) ListNodes() (map[string]*longhorn.Node, error) {
...
+			return nil, err
+		}
+		readyCondition := types.GetNodeConditionFromStatus(node.Status, types.NodeConditionTypeReady)
+		if readyCondition.Status == types.ConditionStatusTrue && node.Spec.AllowScheduling == true {
+			usableNode = node
+			break
+		}
+	}
+	if usableNode == nil {
+		return nil, fmt.Errorf("BUG: Unable to get a ready node")
```
Comment (shuo-wu) [inline datastore/longhorn.go]: Does no available node mean a BUG?

File: controller/volume_controller_test.go [hunk 5]
```diff
@@ -149,7 +167,7 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
 	tc.expectVolume.Status.CurrentImage = tc.volume.Spec.EngineImage
 	testCases["volume detached"] = tc
 
-	// volume attaching, start replicas
+	//volume attaching, start replicas
```
Comment (shuo-wu) [inline controller/volume_controller_test.go]: There is no need to modify this comment. Same below.

File: manager/volume.go [hunk 6]
```diff
@@ -226,16 +226,17 @@ func (m *VolumeManager) Create(name string, spec *types.VolumeSpec) (v *longhorn
...
+			Size:                 size,
+			Frontend:             spec.Frontend,
+			EngineImage:          defaultEngineImage,
+			FromBackup:           spec.FromBackup,
+			NumberOfReplicas:     spec.NumberOfReplicas,
+			StaleReplicaTimeout:  spec.StaleReplicaTimeout,
+			BaseImage:            spec.BaseImage,
+			RecurringJobs:        spec.RecurringJobs,
+			Standby:              spec.Standby,
+			RestorationCompleted: spec.RestorationCompleted,
```
Comment (ywei88) [inline manager/volume.go]: Directly set RestorationCompleted field to false, no need to read from spec

**Comment by yasker** [comment]
Remove the second commit... It's a different PR.

You need to create another work branch locally.

File: controller/volume_controller.go [hunk 7]
```diff
@@ -517,7 +517,6 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, e *longhorn
 	if v.Status.CurrentImage == "" {
 		v.Status.CurrentImage = v.Spec.EngineImage
 	}
-
```
Comment (yasker) [inline controller/volume_controller.go]: Unrelated change.

File: controller/volume_controller.go [hunk 8]
```diff
@@ -528,14 +527,21 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, e *longhorn
 		}
 		newVolume = true
 	}
-
```
Comment (yasker) [inline controller/volume_controller.go]: Unrelated change.

File: controller/volume_controller_test.go [hunk 9]
```diff
@@ -216,6 +235,61 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
 	}
 	testCases["volume attached"] = tc
 
+	// New Restored volume attching to attached
```
Comment (yasker) [inline controller/volume_controller_test.go]: Typo and grammar. `Newly restored volume changed from attaching to attached`

File: controller/volume_controller_test.go [hunk 10]
```diff
@@ -135,7 +135,26 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
 	}
 	testCases["volume create - replica scheduling failure"] = tc
 
-	// after creation, volume in detached state
+	//after creation and detached, restored volume should automatically be in attaching state,
```
Comment (yasker) [inline controller/volume_controller_test.go]: Put the related test cases next to each other in the code.

File: controller/volume_controller_test.go [hunk 11]
```diff
@@ -216,6 +235,61 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
...
+		r.Spec.HealthyAt = getTestNow()
+	}
+	tc.expectVolume.Spec.NodeID = ""
+	tc.expectVolume.Spec.RestorationRequired = false
+	testCases["new restored volume attaching to attached"] = tc
+
+	//volume automatically detaching after restoration completed
+	tc = generateVolumeTestCaseTemplate()
+	tc.volume.Spec.NodeID = ""
+	tc.volume.Spec.FromBackup = "s3://backupbucket@us-east-1/backupstore?backup=backup-4c6353f211e14e39&volume=321"
```
Comment (yasker) [inline controller/volume_controller_test.go]: You've used `random` before? Is the change necessary?

File: controller/volume_controller_test.go [hunk 12]
```diff
@@ -216,6 +235,61 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
...
+	tc.volume.Spec.Standby = false
+	tc.volume.Status.State = types.VolumeStateAttached
+	tc.volume.Spec.RestorationRequired = false
```
Comment (yasker) [inline controller/volume_controller_test.go]: ? Should be `true`?

File: manager/volume.go [hunk 13]
```diff
@@ -178,6 +178,9 @@ func (m *VolumeManager) Create(name string, spec *types.VolumeSpec) (v *longhorn
 		if baseImage, ok := backup.Labels[types.BaseImageLabel]; ok {
 			spec.BaseImage = baseImage
 		}
+		if spec.Standby == false {
+			spec.RestorationRequired = true
```
Comment (yasker) [inline manager/volume.go]: This logic shouldn't be here. It's the internal logic of the controller. Here we simply set volume spec according to the input.

File: controller/volume_controller.go [hunk 14]
```diff
@@ -536,6 +536,23 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, e *longhorn
...
+
+	if neverAttached && v.Spec.FromBackup != "" && v.Spec.Standby == false {
+		v.Spec.RestorationRequired = true
+		usableNode, err := vc.ds.GetRandomReadyNode()
+		if err != nil {
+			return err
+		}
+		v.Spec.NodeID = usableNode.Name
+	}
+
```
Comment (ywei88) [inline controller/volume_controller.go]: The default "RestorationRequired" field is false, and I need to set it to be true for newly restored volume. And all of these logic should be implemented in controller (i.e. not in manager when "create" is called),  so I still need a way to distinguish "newly" restored volume (the restored volume has never been attached), that's the reason for this "neverAttached" boolean.

**Comment by yasker** [comment]
https://github.com/longhorn/longhorn/issues/571
